node/lib/util/commit.js

```diff
@@ -2222,3 +2222,99 @@ it empty. You can remove the commit entirely with "git meta reset HEAD^".`);
                                        message,
                                        subMessages);
 });
+
+const submoduleHasChangesToCommit = function(status, submoduleName) {
+    const submodule = status.submodules[submoduleName];
+    if (!submodule) {
+        return false;
+    }
+    const { workdir } = submodule;
+    return null !== workdir && !workdir.status.isIndexClean();
+};
+
+const submoduleHasPreCommitHook = co.wrap(function *(repo, submoduleName) {
+    const Hook          = require("../util/hook");
+    const SubmoduleUtil = require("../util/submodule_util");
+
+    const subRepo = yield SubmoduleUtil.getRepo(repo, submoduleName);
+    return Hook.hasHook(subRepo, "pre-commit");
+});
+
+const filterInParallel = co.wrap(function *(collection, asyncPredicate) {
+    const DoWorkQueue = require("../util/do_work_queue");
+
+    const booleans = yield DoWorkQueue.doInParallel(collection,
+        asyncPredicate);
+    return collection.filter((item, index) => booleans[index]);
+});
+
+const getSubmodulesNameForPreCommitCheck = co.wrap(function *(repo,
+                                                              repoStatus) {
+    const SubmoduleUtil = require("../util/submodule_util");
+
+    const allSubmoduleNames = yield SubmoduleUtil.getSubmoduleNames(repo);
+    return yield filterInParallel(
+        allSubmoduleNames,
+        co.wrap(function*(submoduleName) {
+            if (!submoduleHasChangesToCommit(repoStatus, submoduleName)) {
+                return false;
+            }
+            return yield submoduleHasPreCommitHook(repo, submoduleName);
+        })
+    );
+});
+
+/**
+ * Execute 'pre-commit' hooks of submodules. Only execute when a submodule
+ * has the hook and has changes to commit.
+ * If there is an error occurred, throw an error
+ * If there are hooks to be run and interactive mode is on, a warning will be
+ * raised to communicate that 'pre-commit' hooks were skipped.
+ *
+ * @async
+ * @param {RepoStatus} status
+ * @param {String}     submoduleName
+ * @return {Boolean}
+ */
+exports.execSubmodulePreCommitHooks = co.wrap(function *(repo,
+                                                         cwd,
+                                                         all,
+                                                         paths,
+                                                         interactive) {
+    const Hook          = require("../util/hook");
+    const SubmoduleUtil = require("../util/submodule_util");
+
+    const repoStatus = yield exports.getCommitStatus(repo, cwd, {
+        all,
+        paths,
+    });
+
+    const submoduleNames = yield getSubmodulesNameForPreCommitCheck(
+        repo,
+        repoStatus
+    );
+
+    if (submoduleNames.length > 0) {
+        if (interactive) {
+            console.warn(
+                "Warning. pre-commit hooks skipped when using option [-i]"
```

Comment:
Could you add a comment here?  Something like "This because pre-commit hooks happen before we offer the commit message for editing, and in interactive mode, we don't know which hooks actually need to be run until the editing is done."

Reply:
Added a comment.